File: backend/app/schemas/grants.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.agent_user_grant import AccessLevel
from app.models.component_grant import ComponentAccessLevel
from app.models.component_access_request import RequestStatus


class ComponentGrantCreate(BaseModel):
    """Schema for granting component access to an agent."""

    component_id: UUID
    agent_id: UUID
    access_level: ComponentAccessLevel = ComponentAccessLevel.VIEWER
    expires_at: Optional[datetime] = None


class ComponentGrantUpdate(BaseModel):
    """Schema for updating a component grant."""

    access_level: Optional[ComponentAccessLevel] = None
    expires_at: Optional[datetime] = None


class ComponentGrantResponse(BaseModel):
    """Response schema for component grants."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    component_id: UUID
    agent_id: UUID
    access_level: ComponentAccessLevel
    granted_by: UUID
    granted_at: datetime
    expires_at: Optional[datetime]
    revoked_at: Optional[datetime]
    is_active: bool


class ComponentGrantListResponse(BaseModel):
    """Schema for paginated component grant list response."""

    data: list[ComponentGrantResponse]
    total: int


class AgentUserGrantCreate(BaseModel):
    """Schema for granting user access to an agent."""

    user_id: UUID
    access_level: AccessLevel
    expires_at: Optional[datetime] = None


class AgentUserGrantResponse(BaseModel):
    """Response schema for agent user grants."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    agent_id: UUID
    user_id: UUID
    access_level: AccessLevel
    granted_by: UUID
    granted_at: datetime
    expires_at: Optional[datetime]
    revoked_at: Optional[datetime]
    is_active: bool


class AgentUserGrantListResponse(BaseModel):
    """Schema for paginated agent user grant list response."""

    data: list[AgentUserGrantResponse]
    total: int


class ComponentAccessRequestCreate(BaseModel):
    """Schema for creating a component access request."""

    component_id: UUID
    agent_id: UUID
    requested_level: ComponentAccessLevel

    def model_post_init(self, __context) -> None:
        """Validate that requested level is not VIEWER (default access)."""
        if self.requested_level == ComponentAccessLevel.VIEWER:
            raise ValueError("Cannot request VIEWER level - this is the default access")


class ComponentAccessRequestResolve(BaseModel):
    """Schema for resolving (approve/deny) an access request."""

    approve: bool
    denial_reason: Optional[str] = None

    def model_post_init(self, __context) -> None:
        """Validate that denial_reason is provided when denying."""
        if not self.approve and not self.denial_reason:
            raise ValueError("denial_reason is required when denying a request")


class ComponentAccessRequestResponse(BaseModel):
    """Response schema for component access requests."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    component_id: UUID
    agent_id: UUID
    agent_name: Optional[str] = None
    requested_level: ComponentAccessLevel
    requested_by: UUID
    requester_name: Optional[str] = None
    requested_at: datetime
    status: RequestStatus
    resolved_by: Optional[UUID]
    resolved_at: Optional[datetime]
    denial_reason: Optional[str]
    is_pending: bool


class ComponentAccessRequestListResponse(BaseModel):
    """Schema for paginated component access request list response."""

    data: list[ComponentAccessRequestResponse]
    total: int

File: frontend/src/types/grants.ts
/**
 * Component access control TypeScript types matching backend schemas.
 */

/**
 * Access level enum for component grants.
 * - viewer: Can view component details (default)
 * - executor: Can execute/use the component
 * - contributor: Can modify the component
 */
export type ComponentAccessLevel = 'viewer' | 'executor' | 'contributor';

/**
 * Request status enum for access requests.
 */
export type RequestStatus = 'pending' | 'approved' | 'denied';

/**
 * Component grant data returned from API.
 */
export interface ComponentGrant {
  id: string;
  component_id: string;
  agent_id: string;
  access_level: ComponentAccessLevel;
  granted_by: string;
  granted_at: string;
  expires_at: string | null;
  revoked_at: string | null;
  is_active: boolean;
}

/**
 * Request body for creating a component grant.
 */
export interface ComponentGrantCreate {
  component_id: string;
  agent_id: string;
  access_level?: ComponentAccessLevel;
  expires_at?: string | null;
}

/**
 * Request body for updating a component grant.
 */
export interface ComponentGrantUpdate {
  access_level?: ComponentAccessLevel;
  expires_at?: string | null;
}

/**
 * Paginated list response for component grants.
 */
export interface ComponentGrantListResponse {
  data: ComponentGrant[];
  total: number;
}

/**
 * Component access request data returned from API.
 */
export interface ComponentAccessRequest {
  id: string;
  component_id: string;
  agent_id: string;
  agent_name: string | null;
  requested_level: ComponentAccessLevel;
  requested_by: string;
  requester_name: string | null;
  requested_at: string;
  status: RequestStatus;
  resolved_by: string | null;
  resolved_at: string | null;
  denial_reason: string | null;
  is_pending: boolean;
}

/**
 * Request body for creating an access request.
 */
export interface ComponentAccessRequestCreate {
  component_id: string;
  agent_id: string;
  requested_level: ComponentAccessLevel;
}

/**
 * Request body for resolving (approve/deny) an access request.
 */
export interface ComponentAccessRequestResolve {
  approve: boolean;
  denial_reason?: string | null;
}

/**
 * Paginated list response for access requests.
 */
export interface ComponentAccessRequestListResponse {
  data: ComponentAccessRequest[];
  total: number;
}

/**
 * Query parameters for listing access requests.
 */
export interface AccessRequestListParams {
  pending_only?: boolean;
  status?: 'pending' | 'approved' | 'denied';
  skip?: number;
  limit?: number;
}
